src/Action.cpp

```diff
@@ -23,60 +19,124 @@ const std::string BaseAction::getStatusString() const
 
-CreateUser::CreateUser(const std::string& name,
-                         const std::string& recommendationType):
-                         name(name), recommendationType(recommendationType) 
-{}
+CreateUser::CreateUser(const std::string& name, const std::string& recommendationAlg): BaseAction(),
+                                                                                        _userName(name),
+                                                                    _recommendationAlg(recommendationAlg) {}
 
-void CreateUser::act(Session& sess)
-{
-    const std::unordered_map<std::string, User*>& users = sess.getUsers();
-    if(users.find(name) != users.end())
-    {
+void CreateUser::act(Session& sess) {
+    if(sess.getUsers().count(_userName)) {
+        error(USER_EXISTS_ERR);
+        return;
+    }
+    if (_recommendationAlg == "len") {
+        sess.addToUserMap(new LengthRecommenderUser(_userName));
+    }
+    else if (_recommendationAlg == "rer") {
+        sess.addToUserMap(new RerunRecommenderUser(_userName));
+    }
+    else if (_recommendationAlg == "gen") {
+        sess.addToUserMap(new GenreRecommenderUser(_userName));
+    }
+    else {
+        error(INVALID_ALG_ERR);
+        return;
+    }
+    complete();
+}
+std::string CreateUser::toString() const {
+    return ("creating new User - " + _userName); //TODO complete due to forum answer
+}
+
+ChangeActiveUser::ChangeActiveUser(const std::string& name): BaseAction(), _newUser(name) {}
 
+void ChangeActiveUser::act(Session &sess) {
+    const std::unordered_map<std::string,User*>& usersMap = sess.getUsers();
+    if(!usersMap.count(_newUser)) {
+        error(USER_DOESNT_EXISTS_ERR);
+    }
+    else {
+        sess.setActiveUser(*(usersMap.at(_newUser)));
+        complete();
     }
 }
 
-void Exit::act(Session& sess)
-{
+std::string ChangeActiveUser::toString() const {
+    return "Active user changed to" + _newUser; //TODO
+}
+
+
+DeleteUser::DeleteUser(const std::string& name): _userName(name) {}
+
+void DeleteUser::act(Session &sess) {
+    if(!sess.getUsers().count(_userName)) {
+        error(USER_DOESNT_EXISTS_ERR);
+    }
+    else {
+        sess.deleteUserFromMap(_userName);
+        complete();
+    }
+}
+
+std::string DeleteUser::toString() const {
+    return ("Deleting user" + _userName); // TODO
+}
+
+
+DuplicateUser::DuplicateUser(const std::string& newUser, const std::string& oldUser): _newUserName(newUser),
+                                                                                      _oldUserName(oldUser){}
+
+void DuplicateUser::act(Session &sess) {
+    const std::unordered_map<std::string,User*>& usersMap = sess.getUsers();
+    if(!usersMap.count(_oldUserName)) {
+        error(USER_DOESNT_EXISTS_ERR);
+    }
+    else if (usersMap.count(_newUserName)) {
+        error(USER_EXISTS_ERR);
+    }
+    else {
+        User& oldUser = *(usersMap.at(_oldUserName));
+        User* newUser = oldUser.clone();
+        sess.addToUserMap(newUser);
```

Comment:
Here I also think that map::find(_oldUserName) would be more efficient

Reply:
same

Reply:
the name is confusing. it doesn't really count... it returns 0 or 1. its basically "contain". find returns an iterator and so less efficient